### PR TITLE
[Metal] Sprinkle [return: Release] on all 'new*' selectors. Fixes #5941.

### LINKED ISSUE #5941: [Metal] Texture disposal results in memory leaks.
### Steps to Reproduce

1. Load an example [project](https://github.com/kav-git/Xamarin-Metal-Issue) someone else created for this a couple of years ago.
2. Replace ``_textures.Add(texture);`` with simple ``_texture.Dispose()`` for simplicity.
3. Build and run.
4. Repeatedly hit the button to invoke texture creation/population (and now disposal per [2] above).

### Expected Behavior

The ability to perform [4] above indefinitely, no monotonic memory usage growth.

### Actual Behavior

Memory usage growth and near term crashes.

### Environment

```
Microsoft Visual Studio Community 2019
Version 16.0.2
VisualStudio.16.Release/16.0.2+28803.202
Microsoft .NET Framework
Version 4.7.03190

Installed Version: Community

Visual C++ 2019   00435-60000-00000-AA908
Microsoft Visual C++ 2019

Application Insights Tools for Visual Studio Package   9.0.20307.1
Application Insights Tools for Visual Studio

ASP.NET and Web Tools 2019   16.0.12311.10635
ASP.NET and Web Tools 2019

ASP.NET Web Frameworks and Tools 2019   16.0.12311.10635
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   16.0.12311.10635
Azure App Service Tools v3.0.0

Azure Functions and Web Jobs Tools   16.0.12311.10635
Azure Functions and Web Jobs Tools

C# Tools   3.0.0-beta4-19170-01+1deafee3682da88bf07d1c18521a99f47446cee8
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Extensibility Message Bus   1.1.77 (master@24013d5)
Provides common messaging-based MEF services for loosely coupled Visual Studio extension components communication and integration.

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 0x10 - v2.9.0.0

Microsoft Continuous Delivery Tools for Visual Studio   0.4
Simplifying the configuration of Azure DevOps pipelines from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft Library Manager   1.0
Install client-side libraries easily to any web project

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

Mono Debugging for Visual Studio   16.0.300 (573eda3)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   5.0.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit http://docs.nuget.org/.

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

ResourcePackage Extension   1.0
ResourcePackage Visual Studio Extension Detailed Info

ResourcePackage Extension   1.0
ResourcePackage Visual Studio Extension Detailed Info

SQL Server Data Tools   16.0.61903.25110
Microsoft SQL Server Data Tools

Test Adapter for Boost.Test   1.0
Enables Visual Studio's testing tools with unit tests written for Boost.Test.  The use terms and Third Party Notices are available in the extension installation directory.

Test Adapter for Google Test   1.0
Enables Visual Studio's testing tools with unit tests written for Google Test.  The use terms and Third Party Notices are available in the extension installation directory.

TypeScript Tools   16.0.10306.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   3.0.0-beta4-19170-01+1deafee3682da88bf07d1c18521a99f47446cee8
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual C++ for Cross Platform Mobile Development (Android)   15.0.28625.91
Visual C++ for Cross Platform Mobile Development (Android)

Visual F# Tools 10.4 for F# 4.6   16.0.0.0.  Commit Hash: 809f41207b74a1356208f11b8fa3a15233734050.
Microsoft Visual F# Tools 10.4 for F# 4.6

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Tools for CMake   1.0
Visual Studio Tools for CMake

Visual Studio Tools for CMake   1.0
Visual Studio Tools for CMake

Visual Studio Tools for Containers   1.0
Visual Studio Tools for Containers

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

Xamarin   16.0.0.515 (d16-0@a3502d929)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   4.17.4.418 (remotes/origin/d16-0@3d086e814)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin Templates   16.1.47 (31f664d)
Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.

Xamarin.Android SDK   9.2.0.5 (HEAD/292d27fa)
Xamarin.Android Reference Assemblies and MSBuild support.
    Mono: mono/mono/2018-08-rc@5ad371dab1b
    Java.Interop: xamarin/java.interop/d16-0@c987483
    LibZipSharp: grendello/LibZipSharp/master@44de300
    LibZip: nih-at/libzip/rel-1-5-1@b95cf3f
    MXE: xamarin/mxe/xamarin@b9cbb535
    ProGuard: xamarin/proguard/master@905836d
    SQLite: xamarin/sqlite/3.26.0@325e91a
    Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-0@0a7edd6


Xamarin.iOS and Xamarin.Mac SDK   12.8.0.0 (0aa8452)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.```

### PR DESCRIPTION
**Pull request by rolfbjarne**
Also add tests for all the API I could figure out how to use.

Fixes https://github.com/xamarin/xamarin-macios/issues/5941.

### CODE REVIEW

File: tests/monotouch-test/Metal/MTLDeviceTests.cs [hunk 1]
```diff
@@ -12,31 +17,418 @@
...
+		}
+		[DllImport (Constants.libcLibrary)]
+		static extern int munmap (IntPtr addr, nint size);
+		static void FreePageAligned (IntPtr ptr, int length)
+		{
+			munmap (ptr, length);
+		}
+
+		[Test]
+		public void ReturnReleaseTest ()
```
Comment (mandel-macaque) [inline tests/monotouch-test/Metal/MTLDeviceTests.cs]: Bravo!

File: tests/monotouch-test/Metal/MTLDeviceTests.cs [hunk 2]
```diff
@@ -12,31 +17,418 @@
...
+
+			using (var queue = device.CreateCommandQueue ()) {
+				using (var scope = MTLCaptureManager.Shared.CreateNewCaptureScope (queue)) {
+					Assert.IsNotNull (scope, "MTLCaptureManager.CreateNewCaptureScope (MTLCommandQueue): nonnull");
+				}
+			}
+
+			TestRuntime.AssertXcodeVersion (10, 0);
+			using (var evt = device.CreateSharedEvent ()) {
+				using (var shared = evt.CreateSharedEventHandle ()) {
```
Comment (mandel-macaque) [inline tests/monotouch-test/Metal/MTLDeviceTests.cs]: Minor, style: no need to use the {} for the first using.

File: tests/monotouch-test/Metal/MTLDeviceTests.cs [hunk 3]
```diff
@@ -12,31 +17,418 @@
...
+						}
+					}
+				}
+			}
+
+			using (var scope = MTLCaptureManager.Shared.CreateNewCaptureScope (device)) {
+				Assert.IsNotNull (scope, "MTLCaptureManager.CreateNewCaptureScope (MTLDevice): nonnull");
+			}
+
+			using (var queue = device.CreateCommandQueue ()) {
```
Comment (mandel-macaque) [inline tests/monotouch-test/Metal/MTLDeviceTests.cs]: Minor, style: no need to use the {} for the first using.

File: tests/monotouch-test/Metal/MTLDeviceTests.cs [hunk 4]
```diff
@@ -12,31 +17,418 @@
...
+				}
+			}
+
+			using (var hd = new MTLHeapDescriptor ()) {
+				hd.CpuCacheMode = MTLCpuCacheMode.DefaultCache;
+				hd.StorageMode = MTLStorageMode.Private;
+				using (var txt = MTLTextureDescriptor.CreateTexture2DDescriptor (MTLPixelFormat.RGBA8Unorm, 40, 40, false)) {
+					var sa = device.GetHeapTextureSizeAndAlign (txt);
+					hd.Size = sa.Size;
+					using (var heap = device.CreateHeap (hd)) {
```
Comment (mandel-macaque) [inline tests/monotouch-test/Metal/MTLDeviceTests.cs]: Minor, style: no need to use the {} for the first using.

File: tests/monotouch-test/Metal/MTLDeviceTests.cs [hunk 5]
```diff
@@ -12,31 +17,418 @@
...
+						Assert.IsNotNull (reflection, "MTLFunction.CreateArgumentEncoder (nuint, MTLArgument): NonNull reflection");
+					}
+				}
+			}
+
+			using (var library = device.CreateDefaultLibrary ()) {
+				using (var func = library.CreateFunction ("grayscaleKernel")) {
+					Assert.IsNotNull (func, "CreateFunction (string): nonnull");
+				}
+				using (var constants = new MTLFunctionConstantValues ()) {
```
Comment (mandel-macaque) [inline tests/monotouch-test/Metal/MTLDeviceTests.cs]: Minor, style: no need to use the {} for the first using.

File: tests/monotouch-test/Metal/MTLDeviceTests.cs [hunk 6]
```diff
@@ -12,31 +17,418 @@
...
+						using (var rps = device.CreateRenderPipelineState (descriptor, MTLPipelineOption.ArgumentInfo, out var reflection, out var error)) {
+							Assert.IsNotNull (rps, "CreateRenderPipelineState (MTLTileRenderPipelineDescriptor, MTLPipelineOption, MTLRenderPipelineReflection, NSError): NonNull");
+							Assert.IsNull (error, "CreateRenderPipelineState (MTLTileRenderPipelineDescriptor, MTLPipelineOption, MTLRenderPipelineReflection, NSError: NonNull error");
+							Assert.IsNotNull (reflection, "CreateRenderPipelineState (MTLTileRenderPipelineDescriptor, MTLPipelineOption, MTLRenderPipelineReflection, NSError): NonNull reflection");
+						}
+					}
+				}
+			}
+
+			using (var buffer = device.CreateBuffer (1024, MTLResourceOptions.CpuCacheModeDefault)) {
```
Comment (mandel-macaque) [inline tests/monotouch-test/Metal/MTLDeviceTests.cs]: Minor, style: no need to use the {} for the first using.

File: tests/monotouch-test/Metal/MTLDeviceTests.cs [hunk 7]
```diff
@@ -12,31 +17,418 @@
...
+				Assert.IsNotNull (evt, "CreateEvent: NonNull");
+			}
+
+			TestRuntime.AssertXcodeVersion (10, 0);
+			using (var evt = device.CreateSharedEvent ()) {
+				Assert.IsNotNull (evt, "CreateSharedEvent: NonNull");
+			}
+
+			TestRuntime.AssertXcodeVersion (10, 0);
+			using (var evt1 = device.CreateSharedEvent ()) {
```
Comment (mandel-macaque) [inline tests/monotouch-test/Metal/MTLDeviceTests.cs]: Minor, style: no need to use the {} for the first and second using.

File: tests/monotouch-test/Metal/MTLDeviceTests.cs [hunk 8]
```diff
@@ -12,31 +17,418 @@
...
+				Assert.IsNull (library, "CreateLibrary (NSUrl, NSError): Null");
+				Assert.IsNotNull (error, "CreateLibrary (NSUrl, NSError): NonNull error");
+			}
+
+			using (var library = device.CreateArgumentEncoder (new MTLArgumentDescriptor [] { new MTLArgumentDescriptor () { DataType = MTLDataType.Int } })) {
+				Assert.IsNotNull (library, "CreateArgumentEncoder (MTLArgumentDescriptor[]): NonNull");
+			}
+
+			TestRuntime.AssertXcodeVersion (10, 0);
+			using (var descriptor = new MTLIndirectCommandBufferDescriptor ()) {
```
Comment (mandel-macaque) [inline tests/monotouch-test/Metal/MTLDeviceTests.cs]: Minor, style: no need to use the {} for the first using.

File: tests/monotouch-test/Metal/MTLDeviceTests.cs [hunk 9]
```diff
@@ -12,31 +17,418 @@
...
+			using (var library = device.CreateDefaultLibrary ()) {
+				using (var func = library.CreateFunction ("grayscaleKernel")) {
+					using (var cps = device.CreateComputePipelineState (func, out var error)) {
+						Assert.IsNotNull (cps, "CreateComputePipelineState: NonNull 2");
+						Assert.IsNull (error, "CreateComputePipelineState: NonNull error 2");
+					}
+				}
+			}
+
+			using (var descriptor = new MTLComputePipelineDescriptor ()) {
```
Comment (mandel-macaque) [inline tests/monotouch-test/Metal/MTLDeviceTests.cs]: Minor, style: I think we can reduce indentation.

File: tests/monotouch-test/Metal/MTLDeviceTests.cs [hunk 10]
```diff
@@ -12,31 +17,418 @@
...
+				using (var func = library.CreateFunction ("grayscaleKernel")) {
+					using (var cps = device.CreateComputePipelineState (func, MTLPipelineOption.ArgumentInfo, out var reflection, out var error)) {
+						Assert.IsNotNull (cps, "CreateComputePipelineState: NonNull 1");
+						Assert.IsNull (error, "CreateComputePipelineState: NonNull error 1");
+						Assert.IsNotNull (reflection, "CreateComputePipelineState: NonNull reflection 1");
+					}
+				}
+			}
+
+			using (var library = device.CreateDefaultLibrary ()) {
```
Comment (mandel-macaque) [inline tests/monotouch-test/Metal/MTLDeviceTests.cs]: Minor, style: I think we can reduce indentation.

File: tests/monotouch-test/Metal/MTLDeviceTests.cs [hunk 11]
```diff
@@ -12,31 +17,418 @@
...
+						using (var rps = device.CreateRenderPipelineState (descriptor, MTLPipelineOption.BufferTypeInfo, out var reflection, out var error)) {
+							Assert.IsNotNull (rps, "CreateRenderPipelineState: NonNull 2");
+							Assert.IsNull (error, "CreateRenderPipelineState: NonNull error 2");
+							Assert.IsNotNull (reflection, "CreateRenderPipelineState: NonNull reflection 2");
+						}
+					}
+				}
+			}
+
+			using (var library = device.CreateDefaultLibrary ()) {
```
Comment (mandel-macaque) [inline tests/monotouch-test/Metal/MTLDeviceTests.cs]: Minor, style: I think we can reduce indentation.

File: tests/monotouch-test/Metal/MTLDeviceTests.cs [hunk 12]
```diff
@@ -12,31 +17,418 @@
...
+						descriptor.ColorAttachments [0].PixelFormat = MTLPixelFormat.BGRA8Unorm_sRGB;
+						using (var rps = device.CreateRenderPipelineState (descriptor, out var error)) {
+							Assert.IsNotNull (rps, "CreateRenderPipelineState: NonNull 1");
+							Assert.IsNull (error, "CreateRenderPipelineState: NonNull error 1");
+						}
+					}
+				}
+			}
+
+			using (var descriptor = new MTLRenderPipelineDescriptor ()) {
```
Comment (mandel-macaque) [inline tests/monotouch-test/Metal/MTLDeviceTests.cs]: Minor, style: I think we can reduce indentation.

File: tests/monotouch-test/Metal/MTLDeviceTests.cs [hunk 13]
```diff
@@ -12,31 +17,418 @@
...
+					Assert.IsNull (error, "CreateLibrary: NonNull error 4");
+				});
+			}
+
+			using (var library = device.CreateDefaultLibrary (NSBundle.MainBundle, out var error)) {
+				Assert.IsNotNull (library, "CreateDefaultLibrary: NonNull 2");
+				Assert.IsNull (error, "CreateDefaultLibrary: NonNull error 2");
+			}
+
+			using (var descriptor = new MTLRenderPipelineDescriptor ()) {
```
Comment (mandel-macaque) [inline tests/monotouch-test/Metal/MTLDeviceTests.cs]: Minor, style: I think we can reduce indentation.

File: tests/monotouch-test/Metal/MTLDeviceTests.cs [hunk 14]
```diff
@@ -12,31 +17,418 @@
...
+			using (var library = device.CreateDefaultLibrary ()) {
+				Assert.IsNotNull (library, "CreateDefaultLibrary: NonNull 1");
+			}
+
+			using (var library = device.CreateLibrary (metallib_path, out var error)) {
+				Assert.IsNotNull (library, "CreateLibrary: NonNull 1");
+				Assert.IsNull (error, "CreateLibrary: NonNull error 1");
+			}
+
+			using (var data = DispatchData.FromByteBuffer (File.ReadAllBytes (metallib_path))) {
```
Comment (mandel-macaque) [inline tests/monotouch-test/Metal/MTLDeviceTests.cs]: Minor, style: I think we can reduce indentation.

File: tests/monotouch-test/Metal/MTLDeviceTests.cs [hunk 15]
```diff
@@ -12,31 +17,418 @@
...
+			var resourceOptions7 = MTLResourceOptions.StorageModeManaged;
+#else
+				var resourceOptions7 = MTLResourceOptions.CpuCacheModeDefault;
+#endif
+			using (var buffer = device.CreateBufferNoCopy (buffer_mem, (nuint)buffer_length, resourceOptions7, (pointer, length) => { FreePageAligned (pointer, (int)length); freed = true; })) {
+				Assert.IsNotNull (buffer, "CreateBufferNoCopy: NonNull 1");
+			}
+			Assert.IsTrue (freed, "CreateBufferNoCopy: Freed 1");
+
+			using (var descriptor = new MTLDepthStencilDescriptor ()) {
```
Comment (mandel-macaque) [inline tests/monotouch-test/Metal/MTLDeviceTests.cs]: Minor, style: no need to use the {} for the first using.

File: tests/monotouch-test/Metal/MTLDeviceTests.cs [hunk 16]
```diff
@@ -12,31 +17,418 @@
...
+				Assert.IsNotNull (queue, "Queue: NonNull 1");
+			}
+
+#if __MACOS__
+			using (var descriptor = MTLTextureDescriptor.CreateTexture2DDescriptor (MTLPixelFormat.RGBA8Unorm, 64, 64, false)) {
+				descriptor.StorageMode = MTLStorageMode.Private;
+				using (var texture = device.CreateSharedTexture (descriptor)) {
+					Assert.IsNotNull (texture, "CreateSharedTexture (MTLTextureDescriptor): NonNull");
+
+					using (var handle = texture.CreateSharedTextureHandle ()) {
```
Comment (mandel-macaque) [inline tests/monotouch-test/Metal/MTLDeviceTests.cs]: Minor, style: no need to use the {} for the first using.

File: tests/monotouch-test/Metal/MTLDeviceTests.cs [hunk 17]
```diff
@@ -12,31 +17,418 @@
...
+			using (var buffer = device.CreateBuffer (buffer_bytes, MTLResourceOptions.CpuCacheModeDefault)) {
+				Assert.IsNotNull (buffer, "CreateBuffer: NonNull 3");
+			}
+
+			buffer_mem = AllocPageAligned (1, out buffer_length);
+			freed = false;
+#if __MACOS__
+			var resourceOptions7 = MTLResourceOptions.StorageModeManaged;
+#else
+				var resourceOptions7 = MTLResourceOptions.CpuCacheModeDefault;
```
Comment (spouliot) [inline tests/monotouch-test/Metal/MTLDeviceTests.cs]: minor: alignment

File: tests/introspection/ApiCtorInitTest.cs [hunk 18]
```diff
@@ -132,6 +132,9 @@ protected virtual bool Skip (Type type)
 				return true;
 			case "QTMovie":
 				return TestRuntime.CheckSystemVersion (PlatformName.MacOSX, 10, 14, 4); // Broke in macOS 10.14.4.
+			case "MTLFunctionConstantValues":
+				// the type was added with Xcode 8, but the default ctor in Xcode 9.
+				return !TestRuntime.CheckXcodeVersion (9, 0);
```
Comment (spouliot) [inline tests/introspection/ApiCtorInitTest.cs]: it would be better to redefine the `.ctor` in the bindings and add the availability attribute

**Comment by spouliot** [comment]
@monojenkins backport to d16-2

**Pull request merged**